File: src/ray/rpc/rpc_chaos.cc
// Copyright 2024 The Ray Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "ray/rpc/rpc_chaos.h"

#include <algorithm>
#include <array>
#include <random>
#include <string>
#include <string_view>
#include <vector>

#include "absl/container/flat_hash_map.h"
#include "absl/synchronization/mutex.h"
#include "nlohmann/json.hpp"
#include "ray/common/ray_config.h"

namespace ray {
namespace rpc {
namespace testing {

// RpcFailureManager is a simple chaos testing framework. Before starting ray, users
// should set up os environment to use this feature for testing purposes.

// You can use this to set probabilities for specific rpc's.
//     export
//     RAY_testing_rpc_failure='{"method1":{"num_failures":3,"req_failure_prob":12,"resp_failure_prob":12,"in_flight_failure_prob":50}}'
//
// You can also use a wildcard to set probabilities for all rpc's and -1 as num_failures
// to have unlimited failures.
//     export
//     RAY_testing_rpc_failure='{"*":{"num_failures":-1,"req_failure_prob":10,"resp_failure_prob":25,"in_flight_failure_prob":50}}'
// This will set the probabilities for all rpc's to 10% for request failures, 25% for
// reply failures, and 50% for in-flight failures.

// You can also provide optional parameters to specify that there
// should be at least a certain amount of request, response, and in-flight failures.
// By default these are set to 0, but by setting them to positive values guarantees that
// the first N RPCs will have X request failures, followed by Y response failures,
// followed by Z in-flight failures. Afterwards, it will revert to the probabilistic
// failures.
//
// You can combine this with the wildcard so that each RPC method will have the same lower
// bounds applied.
//
// Ex. unlimited failures for all rpc's with 25% request failures, 50% response failures,
// and 10% in-flight failures with at least 2 request failures, 3 response failures, and 1
// in-flight failure.
//     export
//     RAY_testing_rpc_failure='{"*":{"num_failures":-1,"req_failure_prob":25,"resp_failure_prob":50,"in_flight_failure_prob":10,"num_lower_bound_req_failures":2,"num_lower_bound_resp_failures":3,"num_lower_bound_in_flight_failures":1}}'

namespace {
constexpr std::array<std::string_view, 7> kValidKeys = {
    "num_failures",
    "req_failure_prob",
    "resp_failure_prob",
    "in_flight_failure_prob",
    "num_lower_bound_req_failures",
    "num_lower_bound_resp_failures",
    "num_lower_bound_in_flight_failures"};
}  // namespace

class RpcFailureManager {
 public:
  RpcFailureManager() { Init(); }

  void Init() {
    absl::MutexLock lock(&mu_);

    // Clear old state
    failable_methods_.clear();
    num_req_failures_.clear();
    num_resp_failures_.clear();
    num_in_flight_failures_.clear();
    wildcard_set_ = false;
    has_failures_ = false;

    if (!RayConfig::instance().testing_rpc_failure().empty()) {
      auto json = nlohmann::json::parse(
          RayConfig::instance().testing_rpc_failure(), nullptr, false);
      if (json.is_discarded()) {
        RAY_LOG(FATAL) << "testing_rpc_failure is not a valid json object: "
                       << RayConfig::instance().testing_rpc_failure();
        return;
      }

      for (const auto &[method, config] : json.items()) {
        if (!config.is_object()) {
          RAY_LOG(FATAL) << "Value for method '" << method
                         << "' in testing_rpc_failure config is not a JSON object: "
                         << config.dump();
          continue;
        }
        for (const auto &[config_key, _] : config.items()) {
          if (std::find(kValidKeys.begin(), kValidKeys.end(), config_key) ==
              kValidKeys.end()) {
            RAY_LOG(FATAL) << "Unknown key specified in testing_rpc_failure config: "
                           << config_key;
          }
        }
        auto [iter, _] = failable_methods_.emplace(
            method,
            Failable{
                // The value method either retrieves the value associated with the key or
                // uses the specified default value instead.
                config.value("num_failures", 0L),
                config.value("req_failure_prob", 0UL),
                config.value("resp_failure_prob", 0UL),
                config.value("in_flight_failure_prob", 0UL),
                config.value("num_lower_bound_req_failures", 0UL),
                config.value("num_lower_bound_resp_failures", 0UL),
                config.value("num_lower_bound_in_flight_failures", 0UL),
            });
        const auto &failable = iter->second;
        RAY_CHECK_LE(failable.req_failure_prob + failable.resp_failure_prob +
                         failable.in_flight_failure_prob,
                     100UL);
        if (method == "*") {
          wildcard_set_ = true;
          // The wildcard overrides all other method configurations.
          break;
        }
      }

      std::random_device rd;
      auto seed = rd();
      RAY_LOG(INFO) << "Setting RpcFailureManager seed to " << seed;
      gen_.seed(seed);
      has_failures_ = true;
    }
  }

  RpcFailure GetRpcFailure(const std::string &name) {
    if (!has_failures_) {
      return RpcFailure::None;
    }

    absl::MutexLock lock(&mu_);

    // Wildcard overrides any other method configurations.
    if (wildcard_set_) {
      return GetFailureTypeFromFailable(failable_methods_["*"], name);
    }

    auto iter = failable_methods_.find(name);
    if (iter == failable_methods_.end()) {
      return RpcFailure::None;
    }
    return GetFailureTypeFromFailable(iter->second, name);
  }

 private:
  absl::Mutex mu_;
  std::mt19937 gen_;
  std::atomic_bool has_failures_ = false;

  // If we're testing all rpc failures, we'll use these probabilites instead of
  // failable_methods_
  bool wildcard_set_ = false;

  struct Failable {
    int64_t num_remaining_failures;
    size_t req_failure_prob;
    size_t resp_failure_prob;
    size_t in_flight_failure_prob;
    size_t num_lower_bound_req_failures = 0;
    size_t num_lower_bound_resp_failures = 0;
    size_t num_lower_bound_in_flight_failures = 0;
  };
  absl::flat_hash_map<std::string, Failable> failable_methods_ ABSL_GUARDED_BY(&mu_);

  absl::flat_hash_map<std::string, size_t> num_req_failures_ ABSL_GUARDED_BY(&mu_);
  absl::flat_hash_map<std::string, size_t> num_resp_failures_ ABSL_GUARDED_BY(&mu_);
  absl::flat_hash_map<std::string, size_t> num_in_flight_failures_ ABSL_GUARDED_BY(&mu_);

  RpcFailure GetFailureTypeFromFailable(Failable &failable, const std::string &name)
      ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
    if (failable.num_remaining_failures == 0) {
      // If < 0, unlimited failures.
      return RpcFailure::None;
    }

    if (num_req_failures_[name] < failable.num_lower_bound_req_failures) {
      failable.num_remaining_failures--;
      num_req_failures_[name]++;
      return RpcFailure::Request;
    }
    if (num_resp_failures_[name] < failable.num_lower_bound_resp_failures) {
      failable.num_remaining_failures--;
      num_resp_failures_[name]++;
      return RpcFailure::Response;
    }
    if (num_in_flight_failures_[name] < failable.num_lower_bound_in_flight_failures) {
      failable.num_remaining_failures--;
      num_in_flight_failures_[name]++;
      return RpcFailure::InFlight;
    }

    std::uniform_int_distribution<size_t> dist(1ul, 100ul);
    const size_t random_number = dist(gen_);
    if (random_number <= failable.req_failure_prob) {
      failable.num_remaining_failures--;
      return RpcFailure::Request;
    } else if (random_number <= failable.req_failure_prob + failable.resp_failure_prob) {
      failable.num_remaining_failures--;
      return RpcFailure::Response;
    } else if (random_number <= failable.req_failure_prob + failable.resp_failure_prob +
                                    failable.in_flight_failure_prob) {
      failable.num_remaining_failures--;
      return RpcFailure::InFlight;
    } else {
      return RpcFailure::None;
    }
  }
};

namespace {

RpcFailureManager &GetRpcFailureManager() {
  static auto *manager = new RpcFailureManager();
  return *manager;
}

}  // namespace

RpcFailure GetRpcFailure(const std::string &name) {
  return GetRpcFailureManager().GetRpcFailure(name);
}

void Init() { GetRpcFailureManager().Init(); }

}  // namespace testing
}  // namespace rpc
}  // namespace ray

File: rust/ray-rpc/src/rpc_chaos.rs
// Copyright 2024 The Ray Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0

//! RPC chaos testing framework.
//!
//! Replaces `src/ray/rpc/rpc_chaos.h/cc`.
//! Provides configurable failure injection for RPC methods, supporting
//! per-method and wildcard configurations with probabilistic and
//! lower-bound guaranteed failure modes.

use std::collections::HashMap;
use std::sync::Mutex;

use rand::Rng;

/// Type of RPC failure to inject.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RpcFailure {
    /// No failure.
    None,
    /// Failure before server receives the request.
    Request,
    /// Failure after server sends the response.
    Response,
    /// Failure after server receives the request but before the response is sent.
    InFlight,
}

/// Valid keys in the JSON configuration for each method.
const VALID_KEYS: &[&str] = &[
    "num_failures",
    "req_failure_prob",
    "resp_failure_prob",
    "in_flight_failure_prob",
    "num_lower_bound_req_failures",
    "num_lower_bound_resp_failures",
    "num_lower_bound_in_flight_failures",
];

/// Configuration for a single failable method.
#[derive(Debug, Clone)]
struct Failable {
    num_remaining_failures: i64,
    req_failure_prob: u64,
    resp_failure_prob: u64,
    in_flight_failure_prob: u64,
    num_lower_bound_req_failures: u64,
    num_lower_bound_resp_failures: u64,
    num_lower_bound_in_flight_failures: u64,
}

/// RPC chaos failure manager.
///
/// Manages failure injection configuration and state for RPC methods.
/// Thread-safe via internal mutex.
pub struct RpcFailureManager {
    inner: Mutex<RpcFailureManagerInner>,
}

struct RpcFailureManagerInner {
    failable_methods: HashMap<String, Failable>,
    wildcard_set: bool,
    has_failures: bool,
    num_req_failures: HashMap<String, u64>,
    num_resp_failures: HashMap<String, u64>,
    num_in_flight_failures: HashMap<String, u64>,
    rng: rand::rngs::ThreadRng,
}

impl RpcFailureManager {
    /// Create a new empty failure manager.
    pub fn new() -> Self {
        Self {
            inner: Mutex::new(RpcFailureManagerInner {
                failable_methods: HashMap::new(),
                wildcard_set: false,
                has_failures: false,
                num_req_failures: HashMap::new(),
                num_resp_failures: HashMap::new(),
                num_in_flight_failures: HashMap::new(),
                rng: rand::thread_rng(),
            }),
        }
    }

    /// Initialize from a JSON configuration string.
    ///
    /// The JSON format is:
    /// ```json
    /// {
    ///   "method_name": {
    ///     "num_failures": 10,
    ///     "req_failure_prob": 25,
    ///     "resp_failure_prob": 25,
    ///     "in_flight_failure_prob": 25,
    ///     "num_lower_bound_req_failures": 0,
    ///     "num_lower_bound_resp_failures": 0,
    ///     "num_lower_bound_in_flight_failures": 0
    ///   }
    /// }
    /// ```
    ///
    /// Use `"*"` as method name for wildcard (applies to all methods).
    /// Use `num_failures: -1` for unlimited failures.
    ///
    /// # Panics
    ///
    /// Panics if the JSON contains unknown keys.
    pub fn init(&self, config_json: &str) {
        let mut inner = self.inner.lock().unwrap();

        // Clear old state.
        inner.failable_methods.clear();
        inner.num_req_failures.clear();
        inner.num_resp_failures.clear();
        inner.num_in_flight_failures.clear();
        inner.wildcard_set = false;
        inner.has_failures = false;

        if config_json.is_empty() {
            return;
        }

        let json: serde_json::Value =
            serde_json::from_str(config_json).expect("testing_rpc_failure is not valid JSON");

        let obj = json.as_object().expect("testing_rpc_failure must be a JSON object");

        for (method, config) in obj {
            let config_obj = config
                .as_object()
                .unwrap_or_else(|| panic!("Value for method '{method}' is not a JSON object"));

            // Validate keys.
            for key in config_obj.keys() {
                if !VALID_KEYS.contains(&key.as_str()) {
                    panic!(
                        "Unknown key specified in testing_rpc_failure config: {key}"
                    );
                }
            }

            let failable = Failable {
                num_remaining_failures: config_obj
                    .get("num_failures")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(0),
                req_failure_prob: config_obj
                    .get("req_failure_prob")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0),
                resp_failure_prob: config_obj
                    .get("resp_failure_prob")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0),
                in_flight_failure_prob: config_obj
                    .get("in_flight_failure_prob")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0),
                num_lower_bound_req_failures: config_obj
                    .get("num_lower_bound_req_failures")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0),
                num_lower_bound_resp_failures: config_obj
                    .get("num_lower_bound_resp_failures")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0),
                num_lower_bound_in_flight_failures: config_obj
                    .get("num_lower_bound_in_flight_failures")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0),
            };

            assert!(
                failable.req_failure_prob
                    + failable.resp_failure_prob
                    + failable.in_flight_failure_prob
                    <= 100,
                "Total failure probability exceeds 100%"
            );

            inner
                .failable_methods
                .insert(method.clone(), failable);

            if method == "*" {
                inner.wildcard_set = true;
                break;
            }
        }

        inner.has_failures = true;
    }

    /// Get the RPC failure type to inject for the given method name.
    pub fn get_rpc_failure(&self, name: &str) -> RpcFailure {
        let mut inner = self.inner.lock().unwrap();

        if !inner.has_failures {
            return RpcFailure::None;
        }

        if inner.wildcard_set {
            // Clone the failable to work around borrow checker.
            let mut failable = inner.failable_methods.get("*").unwrap().clone();
            let result =
                get_failure_type_from_failable(&mut failable, name, &mut inner);
            inner.failable_methods.insert("*".to_string(), failable);
            return result;
        }

        if let Some(mut failable) = inner.failable_methods.get(name).cloned() {
            let result =
                get_failure_type_from_failable(&mut failable, name, &mut inner);
            inner
                .failable_methods
                .insert(name.to_string(), failable);
            result
        } else {
            RpcFailure::None
        }
    }
}

impl Default for RpcFailureManager {
    fn default() -> Self {
        Self::new()
    }
}

fn get_failure_type_from_failable(
    failable: &mut Failable,
    name: &str,
    inner: &mut RpcFailureManagerInner,
) -> RpcFailure {
    if failable.num_remaining_failures == 0 {
        return RpcFailure::None;
    }

    let req_count = *inner.num_req_failures.get(name).unwrap_or(&0);
    if req_count < failable.num_lower_bound_req_failures {
        failable.num_remaining_failures -= 1;
        *inner.num_req_failures.entry(name.to_string()).or_insert(0) += 1;
        return RpcFailure::Request;
    }

    let resp_count = *inner.num_resp_failures.get(name).unwrap_or(&0);
    if resp_count < failable.num_lower_bound_resp_failures {
        failable.num_remaining_failures -= 1;
        *inner.num_resp_failures.entry(name.to_string()).or_insert(0) += 1;
        return RpcFailure::Response;
    }

    let in_flight_count = *inner.num_in_flight_failures.get(name).unwrap_or(&0);
    if in_flight_count < failable.num_lower_bound_in_flight_failures {
        failable.num_remaining_failures -= 1;
        *inner
            .num_in_flight_failures
            .entry(name.to_string())
            .or_insert(0) += 1;
        return RpcFailure::InFlight;
    }

    let random_number: u64 = inner.rng.gen_range(1..=100);

    if random_number <= failable.req_failure_prob {
        failable.num_remaining_failures -= 1;
        RpcFailure::Request
    } else if random_number <= failable.req_failure_prob + failable.resp_failure_prob {
        failable.num_remaining_failures -= 1;
        RpcFailure::Response
    } else if random_number
        <= failable.req_failure_prob
            + failable.resp_failure_prob
            + failable.in_flight_failure_prob
    {
        failable.num_remaining_failures -= 1;
        RpcFailure::InFlight
    } else {
        RpcFailure::None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Port of C++ MethodRpcFailure test.
    /// Tests per-method failure configuration with limited failures.
    #[test]
    fn test_method_rpc_failure() {
        let mgr = RpcFailureManager::new();
        mgr.init(
            r#"{"method1":{"num_failures":0,"req_failure_prob":25,"resp_failure_prob":25,"in_flight_failure_prob":25},"method2":{"num_failures":1,"req_failure_prob":100,"resp_failure_prob":0,"in_flight_failure_prob":0}}"#,
        );

        // Unknown method returns None.
        assert_eq!(mgr.get_rpc_failure("unknown"), RpcFailure::None);

        // method1 has num_failures=0, so always returns None.
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::None);

        // method2 has at most 1 failure with 100% req probability.
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Request);
        // After the single failure is consumed, returns None.
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::None);
    }

    /// Port of C++ MethodRpcFailureEdgeCase test.
    /// Tests that 100% probability for a single failure type always returns that type.
    #[test]
    fn test_method_rpc_failure_edge_case() {
        let mgr = RpcFailureManager::new();
        mgr.init(
            r#"{"method1":{"num_failures":1000,"req_failure_prob":100,"resp_failure_prob":0,"in_flight_failure_prob":0},"method2":{"num_failures":1000,"req_failure_prob":0,"resp_failure_prob":100,"in_flight_failure_prob":0},"method3":{"num_failures":1000,"req_failure_prob":0,"resp_failure_prob":0,"in_flight_failure_prob":100},"method4":{"num_failures":1000,"req_failure_prob":0,"resp_failure_prob":0,"in_flight_failure_prob":0}}"#,
        );

        for _ in 0..1000 {
            assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Request);
            assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Response);
            assert_eq!(mgr.get_rpc_failure("method3"), RpcFailure::InFlight);
            assert_eq!(mgr.get_rpc_failure("method4"), RpcFailure::None);
        }
    }

    /// Port of C++ WildcardRpcFailure test.
    /// Tests wildcard configuration with different failure probabilities.
    #[test]
    fn test_wildcard_rpc_failure() {
        // 100% request failure for all methods.
        let mgr = RpcFailureManager::new();
        mgr.init(
            r#"{"*":{"num_failures":-1,"req_failure_prob":100,"resp_failure_prob":0,"in_flight_failure_prob":0}}"#,
        );
        for _ in 0..100 {
            assert_eq!(mgr.get_rpc_failure("method"), RpcFailure::Request);
        }

        // 100% response failure for all methods.
        let mgr = RpcFailureManager::new();
        mgr.init(
            r#"{"*":{"num_failures":-1,"req_failure_prob":0,"resp_failure_prob":100,"in_flight_failure_prob":0}}"#,
        );
        for _ in 0..100 {
            assert_eq!(mgr.get_rpc_failure("method"), RpcFailure::Response);
        }

        // 100% in-flight failure for all methods.
        let mgr = RpcFailureManager::new();
        mgr.init(
            r#"{"*":{"num_failures":-1,"req_failure_prob":0,"resp_failure_prob":0,"in_flight_failure_prob":100}}"#,
        );
        for _ in 0..100 {
            assert_eq!(mgr.get_rpc_failure("method"), RpcFailure::InFlight);
        }

        // 0% failure probability - all None.
        let mgr = RpcFailureManager::new();
        mgr.init(
            r#"{"*":{"num_failures":-1,"req_failure_prob":0,"resp_failure_prob":0,"in_flight_failure_prob":0}}"#,
        );
        for _ in 0..100 {
            assert_eq!(mgr.get_rpc_failure("method"), RpcFailure::None);
        }
    }

    /// Port of C++ LowerBoundWithWildcard test.
    /// Tests lower bound guaranteed failures followed by probabilistic failures.
    #[test]
    fn test_lower_bound_with_wildcard() {
        let mgr = RpcFailureManager::new();
        mgr.init(
            r#"{
                "*": {
                    "num_failures": -1,
                    "req_failure_prob": 100,
                    "resp_failure_prob": 0,
                    "in_flight_failure_prob": 0,
                    "num_lower_bound_req_failures": 3,
                    "num_lower_bound_resp_failures": 5,
                    "num_lower_bound_in_flight_failures": 2
                }
            }"#,
        );

        // First 3 calls for method1 should be guaranteed Request failures.
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Request);
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Request);
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Request);

        // Next 5 calls should be guaranteed Response failures.
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Response);
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Response);
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Response);
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Response);
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Response);

        // Next 2 calls should be guaranteed InFlight failures.
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::InFlight);
        assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::InFlight);

        // After lower bounds exhausted, 100% request failures (probabilistic).
        for _ in 0..100 {
            assert_eq!(mgr.get_rpc_failure("method1"), RpcFailure::Request);
        }

        // Wildcard applies to any method - method2 should have the same behavior.
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Request);
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Request);
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Request);

        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Response);
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Response);
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Response);
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Response);
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Response);

        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::InFlight);
        assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::InFlight);

        for _ in 0..100 {
            assert_eq!(mgr.get_rpc_failure("method2"), RpcFailure::Request);
        }
    }

    /// Port of C++ TestInvalidJson test.
    /// Tests that unknown keys in config cause a panic.
    #[test]
    #[should_panic(expected = "Unknown key specified in testing_rpc_failure config: invalid_key")]
    fn test_invalid_json() {
        let mgr = RpcFailureManager::new();
        mgr.init(r#"{"*":{"num_failures":-1,"invalid_key":1}}"#);
    }
}
